libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java

```diff
@@ -15,12 +15,16 @@
  * A basic {@link RefCounted} implementation that is initialized with a ref count of 1 and calls {@link #closeInternal()} once it reaches
  * a 0 ref count.
  */
-public abstract class AbstractRefCounted implements RefCounted {
+public abstract class AbstractRefCounted extends AtomicInteger implements RefCounted {
```

Comment:
Not 100% sure about this one, we could just do the improvement to `tryIncRef` here if we feel like we don't want to leak `AtomicInteger` here I. Just figured it was nice to save a little cache here but it's a smaller.

Reply:
Bleurgh. I mean I get why this might be a little better, but still. Could we get the same effect with a `private volatile int` field and a `VarHandle` to do the atomic operations?

Reply:
Yea I guess, it was just laziness to do it this way, let me adjust :)

Reply:
Ah no can't do it because I then lose `updateAndGet` but you know what ... let's leave this out in this step. Let's see how much this matters once the other stuff is optimised. I'll revert this, it's a little extreme :)